File: Python/206. Reverse Linked List.py
# 206. Reverse Linked List.

# Topics: Linked List, Recursion.

'''
## Task:
--------
Given the head of a singly linked list, reverse the list, and return the reversed list.

Example 1:
Input: head = [1,2,3,4,5]
Output: [5,4,3,2,1]

Example 2:
Input: head = [1,2]
Output: [2,1]

Example 3:
Input: head = []
Output: []

Constraints:
The number of nodes in the list is the range [0, 5000].
-5000 <= Node.val <= 5000

Follow up: A linked list can be reversed either iteratively or recursively. Could you implement both?


## Testcase:
------------
[1,2,3,4,5]
[1,2]
[]

## Code:
--------
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
           
'''

# Solution: ---------------------------------------------------------------------------------------------------------

### Solution 1: Iterative Approach with Comments
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev = None  # Initialize previous node as None
        curr = head  # Start with the head as the current node
        while curr:  # Iterate until the end of the list
            next = curr.next  # Save the next node
            curr.next = prev  # Reverse current node's pointer
            prev = curr  # Move prev to the current node
            curr = next  # Move to the next node in the original list
        # At the end, prev will be the new head of the reversed list
        return prev


### Solution 2: Recursive Approach with Comments

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # Base case: if list is empty or has a single node, return head
        if not head or not head.next:
            return head

        # Recursively reverse the rest of the list from the second node
        reversed_list = self.reverseList(head.next)

        # Reverse the link between the current node and the next node
        head.next.next = head
        # Set the current node's next pointer to None to avoid cycles
        head.next = None

        # Return the head of the reversed list
        return reversed_list


# Description: ================================================================================================
'''
To solve the problem of reversing a singly linked list, we can approach it in two ways: iteratively and recursively. 
Here, I'll provide solutions for both methods.

### Iterative Approach
In the iterative approach, we use a loop to reverse the links between nodes. We'll have three pointers: `prev`, `curr`, 
and `next`. Initially, `prev` is set to `None` and `curr` is set to the head of the list. In each iteration, we'll reverse 
the link between the current node and the next node, and then move the pointers one step forward.


### Recursive Approach
In the recursive approach, we dive to the end of the list and then start reversing the links as we backtrack. The base case 
occurs when we reach the end of the list (`head` is `None` or `head.next` is `None`). On backtracking, we set the `next` 
node's `next` pointer to the current node, effectively reversing the link, and then set the current node's `next` pointer 
to `None` to avoid cycles.

In the recursive solution, `reversed_list` holds the head of the reversed list, which is the last node we reach before hitting 
the base case. As we backtrack, we reverse the links and set the `next` pointer of the last node (which becomes the head of the 
reversed list) to `None` to complete the reversal.

'''

### Tests code:

# To test the iterative and recursive solutions for reversing a singly linked list, we can create a small testing framework. 
# This framework will include a function to build a linked list from a Python list, a function to convert a linked list back 
# to a Python list (for easy comparison), and finally, test cases to verify the correctness of both solutions.

# the test framework:
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseListIterative(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev = None
        curr = head
        while curr:
            next = curr.next
            curr.next = prev
            prev = curr
            curr = next
        return prev

    def reverseListRecursive(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head
        reversed_list = self.reverseListRecursive(head.next)
        head.next.next = head
        head.next = None
        return reversed_list

def build_list(elements):
    """Build a linked list from a Python list and return the head."""
    head = current = None
    for element in elements:
        if not head:
            head = current = ListNode(element)
        else:
            current.next = ListNode(element)
            current = current.next
    return head

def to_list(head):
    """Convert a linked list back to a Python list."""
    elements = []
    while head:
        elements.append(head.val)
        head = head.next
    return elements

def test_solution():
    tests = [
        ([1, 2, 3, 4, 5], [5, 4, 3, 2, 1]),
        ([1, 2], [2, 1]),
        ([], []),
    ]
    solution = Solution()
    for input_list, expected_output in tests:
        head = build_list(input_list)
        
        # Test iterative solution
        reversed_head_iterative = solution.reverseListIterative(head)
        result_iterative = to_list(reversed_head_iterative)
        assert result_iterative == expected_output, f"Iterative failed for {input_list}. Expected {expected_output}, got {result_iterative}"

        # Since the list has been reversed, rebuild it for the recursive test
        head = build_list(input_list)

        # Test recursive solution
        reversed_head_recursive = solution.reverseListRecursive(head)
        result_recursive = to_list(reversed_head_recursive)
        assert result_recursive == expected_output, f"Recursive failed for {input_list}. Expected {expected_output}, got {result_recursive}"

    print("All tests passed!")

test_solution()

'''
This test framework defines a few utility functions:
    
- `build_list(elements)` takes a Python list and constructs a linked list.
- `to_list(head)` converts a linked list back into a Python list for easy comparison.
- `test_solution()` runs predefined test cases, including edge cases like an empty list, to ensure both iterative and 
   recursive solutions work as expected.

After running `test_solution()`, it will either print "All tests passed!" if all tests succeed, or it will raise an assertion 
error indicating which test case failed and why. This helps in quickly identifying and fixing any issues in the implementation.

'''

File: Cpp/206. Reverse Linked List.cpp
// 206. Reverse Linked List.

// Topic: Linked List, Recursion.


/*
### Task:
---
Given the head of a singly linked list, reverse the list, and return the reversed list.

Example 1:
Input: head = [1,2,3,4,5]
Output: [5,4,3,2,1]

Example 2:
Input: head = [1,2]
Output: [2,1]

Example 3:
Input: head = []
Output: []
 
Constraints:
The number of nodes in the list is the range [0, 5000].
-5000 <= Node.val <= 5000
 
Follow up: A linked list can be reversed either iteratively or recursively. Could you implement both?

### Testcase:
---
[1,2,3,4,5]
[1,2]
[]


### Code:
---
// *
// * Definition for singly-linked list.
// * struct ListNode {
// *     int val;
// *     ListNode *next;
// *     ListNode() : val(0), next(nullptr) {}
// *     ListNode(int x) : val(x), next(nullptr) {}
// *     ListNode(int x, ListNode *next) : val(x), next(next) {}
// * };
// *
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        
    }
};
*/
// Solution: --------------------------------------

class Solution {
public:
    // Function that matches the test call, using the iterative approach
    ListNode* reverseList(ListNode* head) {
        return reverseListIterative(head);
    }

private:
    // Iterative solution for reversing the linked list
    ListNode* reverseListIterative(ListNode* head) {
        ListNode *prev = nullptr, *curr = head, *next = nullptr;
        while (curr) {
            next = curr->next;  // Store next node
            curr->next = prev;  // Reverse current node's pointer
            prev = curr;        // Move pointers one position ahead
            curr = next;
        }
        return prev;  // New head of the reversed list
    }

    // Recursive solution for reversing the linked list
    ListNode* reverseListRecursive(ListNode* head) {
        if (!head || !head->next) return head; // Base case: empty list or single node
        ListNode* restReversed = reverseListRecursive(head->next); // Recursively reverse the rest of the list
        head->next->next = head; // Set next node's next to current node
        head->next = nullptr;    // Set current node's next to nullptr
        return restReversed;     // Return the new head of the reversed list
    }
};



// Description: ===================================
/*
This solution provides two approaches to reverse a singly linked list: iterative and recursive. Both methods aim to rewire the `next` pointers of the nodes so that they point to their previous nodes, effectively reversing the direction of the list.

### Iterative Approach:
- **Initialization**: Start with three pointers: `prev` (initially `nullptr`), `curr` (pointing to the head of the list), and `next` (used to temporarily store the next node).
- **Traversal and Reversal**: Iterate through the list with `curr`, at each step:
  - Temporarily store the next node (`next = curr->next`).
  - Reverse the current node's pointer (`curr->next = prev`).
  - Move `prev` and `curr` pointers one step forward.
- **Termination**: When `curr` becomes `nullptr`, `prev` will be pointing at the new head of the reversed list, which is then returned.

### Recursive Approach:
- **Base Case**: If the list is empty (`head == nullptr`) or has only one node (`head->next == nullptr`), return `head` as it is already "reversed".
- **Recursive Step**: Assume the rest of the list is reversed, and you are left with the task of adding the initial head to the end of the reversed list.
  - Recursively reverse the rest of the list starting from the second node (`reverseListRecursive(head->next)`).
  - Make the next node of the original head (`head->next`, which is now the last node of the reversed sublist) point back to the head.
  - Set `head->next` to `nullptr` to break the original next link, making the original head the new tail.
- **Termination**: Return the head of the reversed list obtained from the recursive call, which bubbles up as the final result.

### Implementation:
The provided code defines a `Solution` class with three public methods: `reverseList`, `reverseListIterative`, and `reverseListRecursive`. The `reverseList` method is designed to be the primary interface, which internally calls the iterative approach (`reverseListIterative`). However, it can be easily modified to use the recursive approach by changing its implementation to call `reverseListRecursive`.

### Usage:
- The iterative method is straightforward and generally more efficient in terms of space usage as it does not incur the overhead of recursive calls.
- The recursive method, while elegant and a good demonstration of the divide-and-conquer strategy, uses additional stack space for recursive calls, which could be a concern for very long lists.

Both methods achieve the reversal of the linked list with linear time complexity, \( O(n) \), where \( n \) is the number of nodes in the list. The space complexity of the iterative method is \( O(1) \), while the recursive method has a space complexity of \( O(n) \) due to the recursive call stack.

*/
